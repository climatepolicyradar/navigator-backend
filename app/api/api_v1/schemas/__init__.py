import re

CLIMATE_LAWS_MATCH = re.compile(r"^https?://(.+\.)?climate-laws.org(/|$)")
