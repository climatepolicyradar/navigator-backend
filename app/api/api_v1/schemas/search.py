from enum import Enum
from typing import Mapping, Optional, Sequence

from pydantic import BaseModel, conlist, validator

from app.db.models.law_policy import FamilyCategory
from . import CLIMATE_LAWS_MATCH


Coord = tuple[float, float]


class SortOrder(str, Enum):
    """Sort ordering for use building OpenSearch query body."""

    ASCENDING = "asc"
    DESCENDING = "desc"


class SortField(str, Enum):
    """Sort field for use building OpenSearch query body."""

    DATE = "date"
    TITLE = "title"


class JitQuery(str, Enum):
    """Flag used for determining if a jit query is to be used."""

    ENABLED = "enabled"
    DISABLED = "disabled"


class FilterField(str, Enum):
    """Filter field for use building OpenSearch query body."""

    SOURCE = "sources"
    COUNTRY = "countries"
    REGION = "regions"
    CATEGORY = "categories"
    LANGUAGE = "languages"


class IncludedResults(str, Enum):
    """Filter field to exclude specific results from search based on search indices."""

    PDFS_TRANSLATED = "pdfsTranslated"
    HTMLS_NON_TRANSLATED = "htmlsNonTranslated"
    HTMLS_TRANSLATED = "htmlsTranslated"


IncludedResultsList = Optional[conlist(IncludedResults, min_items=1)]


class SearchRequestBody(BaseModel):
    """The request body expected by the search API endpoint."""

    query_string: str
    exact_match: bool = False
    max_passages_per_doc: int = 10  # TODO: decide on default

    # TODO: Improve filters to allow generics & use filter types
    keyword_filters: Optional[Mapping[FilterField, Sequence[str]]] = None
    year_range: Optional[tuple[Optional[int], Optional[int]]] = None

    sort_field: Optional[SortField] = None
    sort_order: SortOrder = SortOrder.DESCENDING

    include_results: IncludedResultsList = None

    limit: int = 10  # TODO: decide on default
    offset: int = 0

    continuation_token: Optional[str] = None


class SearchResponseDocumentPassage(BaseModel):
    """A Document passage match returned by the search API endpoint."""

    text: str
    text_block_id: str
    text_block_page: Optional[int]
    text_block_coords: Optional[Sequence[Coord]]


class OpenSearchResponseMatchBase(BaseModel):
    """Describes matches returned by an OpenSearch query"""

    document_name: str
    document_geography: str
    document_description: str
    document_sectors: Sequence[str]
    document_source: str
    document_id: str  # Changed semantics to be import_id, not database id
    document_date: str
    document_type: str
    document_source_url: Optional[str]
    document_cdn_object: Optional[str]
    document_category: str
    document_content_type: Optional[str]
    document_slug: str


class OpenSearchResponseNameMatch(OpenSearchResponseMatchBase):
    """Describes matches returned by OpenSearch on Document name."""

    for_search_document_name: str


class OpenSearchResponseDescriptionMatch(OpenSearchResponseMatchBase):
    """Describes matches returned by OpenSearch on Document description."""

    for_search_document_description: str


class OpenSearchResponsePassageMatch(OpenSearchResponseMatchBase):
    """Describes matches returned by OpenSearch on Document passage."""

    text: str
    text_block_id: str
    text_block_page: Optional[int]
    text_block_coords: Optional[Sequence[Coord]]

    @validator("text_block_page", always=True)
    @classmethod
    def validate_page(cls, value):
        """PDF page numbers must be incremented from our 0-indexed values."""
        if value is None:
            return None
        return value + 1


class SearchResponseFamilyDocument(BaseModel):
    """A single document in a search response."""

    document_title: str
    document_slug: str
    document_type: str
    document_source_url: Optional[str]
    document_url: Optional[str]
    document_content_type: Optional[str]
    document_passage_matches: list[SearchResponseDocumentPassage]

    @validator("document_source_url")
    def _filter_climate_laws_url_from_source(cls, v):
        """Make sure we do not return climate-laws.org source URLs to the frontend"""
        if v is None or CLIMATE_LAWS_MATCH.match(v) is not None:
            return None
        return v


class SearchResponseFamily(BaseModel):
    """
    The object that is returned in the response.

    Used to extend with postfix
    """

    family_slug: str
    family_name: str
    family_description: str
    family_category: str
    family_date: str
    family_last_updated_date: str
    family_source: str
    family_geography: str
    family_metadata: dict
    family_title_match: bool
    family_description_match: bool
    family_documents: list[SearchResponseFamilyDocument]


class SearchResponse(BaseModel):
    """The response body produced by the search API endpoint."""

    hits: int
    query_time_ms: int
    total_time_ms: int
    continuation_token: Optional[str] = None

    families: Sequence[SearchResponseFamily]


Top5FamilyList = conlist(SearchResponseFamily, max_items=5)
# Alias required for type hinting
_T5FamL = Top5FamilyList


class GeographySummaryFamilyResponse(BaseModel):
    """Additional information for the geography page over geo stats"""

    family_counts: Mapping[FamilyCategory, int]
    top_families: Mapping[FamilyCategory, _T5FamL]
    targets: Sequence[str]  # TODO: Placeholder for later
