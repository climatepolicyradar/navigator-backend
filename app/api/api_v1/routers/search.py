"""
Searches for documents.

All endpoints should perform document searches using the SearchRequestBody as
its input. The individual endpoints will return different responses tailored
for the type of document search being performed.
"""

import logging
from io import BytesIO
from typing import Annotated, Optional, Sequence, cast

from cpr_sdk.exceptions import QueryError
from cpr_sdk.search_adaptors import VespaSearchAdapter
from fastapi import APIRouter, Body, Depends, Header, HTTPException, Request, status
from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session
from starlette.responses import RedirectResponse

from app.api.api_v1.schemas.search import SearchRequestBody, SearchResponse
from app.core.aws import S3Client, S3Document, get_s3_client
from app.core.config import (
    AWS_REGION,
    CDN_DOMAIN,
    DOC_CACHE_BUCKET,
    INGEST_TRIGGER_ROOT,
    PIPELINE_BUCKET,
    PUBLIC_APP_URL,
    VESPA_SECRETS_LOCATION,
    VESPA_URL,
)
from app.core.custom_app import AppTokenFactory
from app.core.download import create_data_download_zip_archive
from app.core.search import (
    create_vespa_search_params,
    process_result_into_csv,
    process_vespa_search_response,
)
from app.db.session import get_db

_LOGGER = logging.getLogger(__name__)

_VESPA_CONNECTION = VespaSearchAdapter(
    instance_url=VESPA_URL,
    cert_directory=VESPA_SECRETS_LOCATION,
)


search_router = APIRouter()


def _search_request(db: Session, search_body: SearchRequestBody) -> SearchResponse:
    is_browse_request = not search_body.query_string
    if is_browse_request:
        search_body.all_results = True
        search_body.documents_only = True
        search_body.exact_match = False
    try:
        cpr_sdk_search_params = create_vespa_search_params(db, search_body)
        cpr_sdk_search_response = _VESPA_CONNECTION.search(
            parameters=cpr_sdk_search_params
        )
    except QueryError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid Query: {' '.join(e.args)}",
        )
    return process_vespa_search_response(
        db,
        cpr_sdk_search_response,
        limit=search_body.page_size,
        offset=search_body.offset,
    ).increment_pages()


@search_router.post("/searches")
def search_documents(
    request: Request,
    search_body: Annotated[
        SearchRequestBody,
        Body(
            openapi_examples={
                "simple": {
                    "summary": "Simple Search Example",
                    "description": "Perform a simple search for matching passages",
                    "value": {
                        "query_string": "Energy Prices",
                    },
                },
                "browse request": {
                    "summary": "Browse Example",
                    "description": "Perform a browse request within a year range",
                    "value": {
                        "query_string": "",
                        "year_range": [2000, None],
                        "sort_field": "date",
                        "sort_order": "desc",
                    },
                },
                "filters": {
                    "summary": "Filter example",
                    "description": "Filtering and using exact match",
                    "value": {
                        "query_string": "Just transition",
                        "exact_match": True,
                        "keyword_filters": {
                            "sources": ["CCLW"],
                            "categories": ["Legislative"],
                        },
                    },
                },
            }
        ),
    ],
    app_token: Annotated[str, Header()],
    db=Depends(get_db),
) -> SearchResponse:
    """
    Search for documents matching the search criteria and filters.

    There is no authentication required for using this interface. We ask that users be
    respectful of its use and remind users that data is available to download on
    request.

    The search endpoint behaves in two distinct ways:
        - "Browse" mode is when an empty `query_string` is provided. This is intended
        for document level search using filters. Individual passages are not returned.
        - "Search" mode is when a `query_string` is present. This matches against
        individual document passages.

    The request and response object is otherwise identical for both.

    The results can be paginated via a combination of limit, offset and continuation
    tokens. The limit/offset slices the results after they have been retrieved from
    the search database. The continuation token can be used to get the next set of
    results from the search database. See the request schema for more details.
    """
    _LOGGER.info(
        "Search request",
        extra={
            "props": {
                "search_request": search_body.model_dump(),
                "app_token": str(app_token),
            }
        },
    )

    # Decode the app token and validate it.
    #
    # First corpora validation is app token against DB. At least one of the app token
    # corpora IDs must be present in the DB to continue the search request.
    token = AppTokenFactory()
    token.decode_and_validate(db, request, app_token)

    # If the search request IDs are null, we want to search using the app token corpora.
    if search_body.corpus_import_ids is None:
        search_body.corpus_import_ids = cast(Sequence, token.allowed_corpora_ids)

    # For the second validation, search request corpora Ids are validated against the
    # app token corpora IDs if the search request param 'corpus_import_ids' is not None.
    # corpus_import_ids must be a subset of app token IDs.
    token.validate_subset(
        set(search_body.corpus_import_ids), cast(set, token.allowed_corpora_ids)
    )

    _LOGGER.info(
        "Starting search...",
        extra={"props": {"search_request": search_body.model_dump()}},
    )
    return _search_request(db=db, search_body=search_body)


@search_router.post("/searches/download-csv", include_in_schema=False)
def download_search_documents(
    request: Request,
    search_body: SearchRequestBody,
    app_token: Annotated[str, Header()],
    db=Depends(get_db),
) -> StreamingResponse:
    """Download a CSV containing details of documents matching the search criteria."""
    token = AppTokenFactory()

    _LOGGER.info(
        "Search download request",
        extra={
            "props": {
                "search_request": search_body.model_dump(),
                "app_token": str(app_token),
            }
        },
    )

    # Decode the app token and validate it.
    #
    # First corpora validation is app token against DB. At least one of the app token
    # corpora IDs must be present in the DB to continue the search request.
    token = AppTokenFactory()
    token.decode_and_validate(db, request, app_token)

    # If the search request IDs are null, we want to search using the app token corpora.
    if search_body.corpus_import_ids is None:
        search_body.corpus_import_ids = cast(Sequence, token.allowed_corpora_ids)

    # For the second validation, search request corpora Ids are validated against the
    # app token corpora IDs if the search request param 'corpus_import_ids' is not None.
    # corpus_import_ids must be a subset of app token IDs.
    token.validate_subset(
        set(search_body.corpus_import_ids), cast(set, token.allowed_corpora_ids)
    )

    is_browse = not bool(search_body.query_string)

    _LOGGER.info(
        "Starting search...",
        extra={"props": {"search_request": search_body.model_dump()}},
    )
    search_response = _search_request(
        db=db,
        search_body=search_body,
    )
    content_str = process_result_into_csv(db, search_response, is_browse=is_browse)

    _LOGGER.debug(f"Downloading search results as CSV: {content_str}")
    return StreamingResponse(
        content=BytesIO(content_str.encode("utf-8")),
        headers={
            "Content-Type": "text/csv",
            "Content-Disposition": "attachment; filename=results.csv",
        },
    )


def _get_s3_doc_url_from_cdn(
    s3_client: S3Client, s3_document: S3Document, data_dump_s3_key: str
) -> Optional[str]:
    redirect_url = None
    if s3_client.document_exists(s3_document):
        _LOGGER.info("Redirecting to CDN data dump location...")
        redirect_url = f"https://{CDN_DOMAIN}/{data_dump_s3_key}"
    return redirect_url


@search_router.get("/searches/download-all-data", include_in_schema=False)
def download_all_search_documents(
    request: Request, app_token: Annotated[str, Header()], db=Depends(get_db)
) -> RedirectResponse:
    """Download a CSV containing details of all the documents in the corpus."""
    token = AppTokenFactory()

    _LOGGER.info(
        "Whole data download request",
        extra={
            "props": {
                "app_token": str(app_token),
            }
        },
    )

    # Decode the app token and validate it.
    token = AppTokenFactory()
    token.decode_and_validate(db, request, app_token)

    if PIPELINE_BUCKET is None or PUBLIC_APP_URL is None or DOC_CACHE_BUCKET is None:
        if PIPELINE_BUCKET is None:
            _LOGGER.error("{PIPELINE_BUCKET} is not set")
        if PUBLIC_APP_URL is None:
            _LOGGER.error("{PUBLIC_APP_URL} is not set")
        if DOC_CACHE_BUCKET is None:
            _LOGGER.error("{DOC_CACHE_BUCKET} is not set")
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Missing required environment variables",
        )

    s3_client = get_s3_client()
    latest_ingest_start = s3_client.get_latest_ingest_start(
        PIPELINE_BUCKET, INGEST_TRIGGER_ROOT
    )

    s3_prefix = "navigator/dumps"
    data_dump_s3_key = (
        f"{s3_prefix}/{token.sub}-whole_data_dump-{latest_ingest_start}.zip"
    )

    valid_credentials = s3_client.is_connected()
    if not valid_credentials:
        _LOGGER.info("Error connecting to S3 AWS")
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED, detail="Error connecting to AWS"
        )

    s3_document = S3Document(DOC_CACHE_BUCKET, AWS_REGION, data_dump_s3_key)
    if valid_credentials is True and (not s3_client.document_exists(s3_document)):
        aws_env = "production" if "dev" not in PUBLIC_APP_URL else "staging"
        _LOGGER.info(
            f"Generating {token.sub} {aws_env} dump for ingest cycle w/c {latest_ingest_start}..."
        )

        # After writing to a file buffer the position stays at the end whereas when you
        # upload a buffer, it starts from the position it is currently in. We need to
        # add the seek(0) to reset the buffer position to the beginning before writing
        # to S3 to avoid creating an empty file.
        #
        # FIXME Add where clause to filter out MCF data
        zip_buffer = create_data_download_zip_archive(
            latest_ingest_start, token.allowed_corpora_ids, db
        )
        zip_buffer.seek(0)

        try:
            response = s3_client.upload_fileobj(
                bucket=DOC_CACHE_BUCKET,
                key=data_dump_s3_key,
                content_type="application/zip",
                fileobj=zip_buffer,
            )
            if response is False:
                _LOGGER.error("Failed to upload archive to s3: %s", response)
            else:
                _LOGGER.info(f"Finished uploading data archive to {DOC_CACHE_BUCKET}")

        except Exception as e:
            _LOGGER.error(e)

    s3_document = S3Document(DOC_CACHE_BUCKET, AWS_REGION, data_dump_s3_key)
    redirect_url = _get_s3_doc_url_from_cdn(s3_client, s3_document, data_dump_s3_key)
    if redirect_url is not None:
        return RedirectResponse(redirect_url, status_code=status.HTTP_303_SEE_OTHER)

    _LOGGER.info(
        f"Can't find data dump for {latest_ingest_start} in {DOC_CACHE_BUCKET}"
    )
    raise HTTPException(status_code=status.HTTP_404_NOT_FOUND)
