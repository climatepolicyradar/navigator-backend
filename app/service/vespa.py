"""Adaptors for searching CPR data"""

import logging
import time
from abc import ABC
from pathlib import Path
from typing import Optional

from cpr_sdk.exceptions import DocumentNotFoundError, FetchError, QueryError
from cpr_sdk.models.search import Hit, SearchParameters, SearchResponse
from cpr_sdk.vespa import (
    VespaErrorDetails,
    build_vespa_request_body,
    find_vespa_cert_paths,
    parse_vespa_response,
    split_document_id,
)
from requests.exceptions import HTTPError
from vespa.application import Vespa, VespaAsync
from vespa.exceptions import VespaError

LOGGER = logging.getLogger(__name__)


class SearchAdapter(ABC):
    """Base class for all search adapters."""

    def search(self, parameters: SearchParameters) -> SearchResponse:
        """
        Search a dataset

        :param SearchParameters parameters: a search request object
        :return SearchResponse: a list of parent families, each containing relevant
            child documents and passages
        """
        raise NotImplementedError

    def get_by_id(self, document_id: str) -> SearchResponse:
        """
        Get a single document by its id

        :param str document_id: document id
        :return Hit: a single document or passage
        """
        raise NotImplementedError


class VespaAsyncSearchAdapter(SearchAdapter):
    """Search within a Vespa instance."""

    instance_url: str
    client: VespaAsync

    def __init__(
        self,
        instance_url: str,
        cert_directory: Optional[str] = None,
    ):
        """
        Initialize the Vespa search adapter.

        :param instance_url: URL of the Vespa instance to connect to
        :param cert_directory: Optional directory containing cert.pem and key.pem files.
            If None, will attempt to find certs automatically.
        """
        self.instance_url = instance_url
        if cert_directory is None:
            cert_path, key_path = find_vespa_cert_paths()
        else:
            cert_path = (Path(cert_directory) / "cert.pem").__str__()
            key_path = (Path(cert_directory) / "key.pem").__str__()

        self.client = VespaAsync(
            app=Vespa(url=instance_url, cert=cert_path, key=key_path)
        )

    def search(self, parameters: SearchParameters) -> SearchResponse:
        """
        Search a vespa instance

        :param SearchParameters parameters: a search request object
        :return SearchResponse: a list of families, with response metadata
        """
        total_time_start = time.time()
        vespa_request_body = build_vespa_request_body(parameters)
        query_time_start = time.time()
        try:
            vespa_response = self.client.query(body=vespa_request_body)  # type: ignore
        except VespaError as e:
            err_details = VespaErrorDetails(e)
            if err_details.is_invalid_query_parameter:
                LOGGER.error(err_details.message)
                raise QueryError(err_details.summary)
            else:
                raise e
        query_time_end = time.time()

        response = parse_vespa_response(vespa_response=vespa_response)  # type: ignore

        response.query_time_ms = int((query_time_end - query_time_start) * 1000)
        response.total_time_ms = int((time.time() - total_time_start) * 1000)

        return response

    def get_by_id(self, document_id: str) -> Hit:  # type: ignore
        """
        Get a single document by its id

        :param str document_id: IDs should look something like
            "id:doc_search:family_document::CCLW.family.11171.0"
            or
            "id:doc_search:document_passage::UNFCCC.party.1060.0.3743"
        :return Hit: a single document or passage
        """
        namespace, schema, data_id = split_document_id(document_id)
        try:
            vespa_response = self.client.get_data(  # type: ignore
                namespace=namespace, schema=schema, data_id=data_id
            )
        except HTTPError as e:
            if e.response is not None:
                status_code = e.response.status_code
            else:
                status_code = "Unknown"
            if status_code == 404:
                raise DocumentNotFoundError(document_id) from e
            else:
                raise FetchError(
                    f"Received status code {status_code} when fetching "
                    f"document {document_id}",
                    status_code=status_code,
                ) from e

        return Hit.from_vespa_response(vespa_response.json)  # type: ignore
