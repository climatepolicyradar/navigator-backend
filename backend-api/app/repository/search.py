"""Functions to support browsing the RDS document structure"""

from datetime import datetime
from logging import getLogger
from time import perf_counter_ns
from typing import cast

from db_client.models.dfce.family import (
    Corpus,
    DocumentStatus,
    Family,
    FamilyCorpus,
    FamilyDocument,
)
from db_client.models.organisation import Organisation
from sqlalchemy.orm import Session

from app.models.search import (
    BrowseArgs,
    SearchResponse,
    SearchResponseFamily,
    SortField,
    SortOrder,
)
from app.repository.geography import get_geo_subquery
from app.telemetry import observe

_LOGGER = getLogger(__name__)


@observe(name="to_search_response_family")
def to_search_response_family(
    family: Family,
    corpus: Corpus,
    geography_value: str,
    organisation: Organisation,
) -> SearchResponseFamily:
    family_published_date = ""
    if family.published_date is not None:
        family_published_date = family.published_date.isoformat()

    family_last_updated_date = ""
    if family.last_updated_date is not None:
        family_last_updated_date = family.last_updated_date.isoformat()

    return SearchResponseFamily(
        family_slug=cast(str, family.slugs[-1].name),
        family_name=cast(str, family.title),
        family_description=cast(str, family.description),
        family_category=str(family.family_category),
        family_date=family_published_date,
        family_last_updated_date=family_last_updated_date,
        family_source=cast(str, organisation.name),
        corpus_import_id=cast(str, corpus.import_id),
        corpus_type_name=cast(str, corpus.corpus_type_name),
        family_geographies=[row.value for row in family.geographies],
        family_title_match=False,
        family_description_match=False,
        # â†“ Stuff we don't currently use for browse â†“
        total_passage_hits=0,
        family_metadata={},
        family_documents=[],
    )


@observe(name="browse_rds_families")
def browse_rds_families(db: Session, req: BrowseArgs) -> tuple[int, SearchResponse]:
    """Browse RDS"""

    t0 = perf_counter_ns()
    geo_subquery = get_geo_subquery(db, req.geography_slugs, req.country_codes)
    # Subquery to find families with at least one published document
    # Avoid using calculated family_status field
    published_families = (
        db.query(FamilyDocument.family_import_id)
        .filter(FamilyDocument.document_status == DocumentStatus.PUBLISHED)
        .distinct()
        .subquery()
    )
    query = (
        db.query(Family, Corpus, geo_subquery.c.value, Organisation)  # type: ignore
        .join(FamilyCorpus, FamilyCorpus.family_import_id == Family.import_id)
        .join(Corpus, FamilyCorpus.corpus_import_id == Corpus.import_id)
        .join(Organisation, Organisation.id == Corpus.organisation_id)
        .join(
            published_families,
            published_families.c.family_import_id == Family.import_id,
        )
        .filter(geo_subquery.c.family_import_id == Family.import_id)  # type: ignore
    )

    if req.categories is not None:
        query = query.filter(Family.family_category.in_(req.categories))

    if req.corpora_ids is not None and req.corpora_ids != []:
        query = query.filter(Corpus.import_id.in_(req.corpora_ids))

    if req.sort_field == SortField.TITLE:
        if req.sort_order == SortOrder.DESCENDING:
            query = query.order_by(Family.title.desc())
        else:
            query = query.order_by(Family.title.asc())

    _LOGGER.debug("Starting families query")
    families_count = query.count()
    top_five_families = query.limit(5).all()
    families = [
        to_search_response_family(family, corpus, geography_value, organisation)
        for (family, corpus, geography_value, organisation) in top_five_families
    ]

    _LOGGER.debug("Finished families query")

    # Dates are calculated, and therefore sorting cannot be implemented in the query
    if req.start_year is not None:
        compare_date = datetime(year=req.start_year, month=1, day=1).isoformat()
        families = list(
            filter(
                lambda f: f.family_date != "" and f.family_date >= compare_date,
                families,
            )
        )

    if req.end_year is not None:
        compare_date = datetime(year=req.end_year, month=12, day=31).isoformat()
        families = list(
            filter(
                lambda f: f.family_date != "" and f.family_date <= compare_date,
                families,
            )
        )

    if req.sort_field == SortField.DATE:
        families = sorted(
            list(filter(lambda f: f.family_date != "", families)),
            key=lambda f: f.family_date,
            reverse=req.sort_order == SortOrder.DESCENDING,
        ) + list(filter(lambda f: f.family_date == "", families))

    offset = req.offset or 0
    limit = req.limit or len(families)
    time_taken = int((perf_counter_ns() - t0) / 1e6)

    return (
        families_count,
        SearchResponse(
            hits=families_count,
            total_family_hits=families_count,
            query_time_ms=time_taken,
            total_time_ms=time_taken,
            families=families[offset : offset + limit],
        ),
    )
