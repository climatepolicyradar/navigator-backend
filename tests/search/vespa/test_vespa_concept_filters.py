from unittest.mock import patch

import pytest

from app.api.api_v1.routers import search
from app.api.api_v1.schemas.search import SearchResponse
from tests.search.vespa.setup_search_tests import (
    _make_search_request,
    _populate_db_families,
)


@pytest.mark.search
@patch("app.api.api_v1.routers.search.verify_any_corpora_ids_in_db", return_value=True)
@pytest.mark.parametrize(
    "label,query,concept_filters",
    [
        ("search", "the", [{"name": "name", "value": "environment"}]),
        (
            "search",
            "the",
            [
                {"name": "parent_concept_ids_flat", "value": "Q0,"},
                {"name": "id", "value": "0"},
            ],
        ),
        (
            "browse",
            "",
            [
                {"name": "model", "value": "sectors_1"},
                {"name": "id", "value": "0"},
            ],
        ),
    ],
)
def test_concept_filters(
    mock_corpora_exist_in_db,
    label,
    query,
    concept_filters,
    test_vespa,
    data_db,
    monkeypatch,
    data_client,
    valid_token,
):
    monkeypatch.setattr(search, "_VESPA_CONNECTION", test_vespa)

    _populate_db_families(data_db, deterministic_metadata=True)

    response = _make_search_request(
        data_client,
        valid_token,
        {
            "query_string": query,
            "concept_filters": concept_filters,
        },
        expected_status_code=200,
    )
    response: SearchResponse = SearchResponse.model_validate(response)

    assert len(response.families) > 0
    for concept_filter in concept_filters:
        for family in response.families:
            for family_document in family.family_documents:
                for passage_match in family_document.document_passage_matches:
                    assert passage_match.concepts
                    if concept_filter["name"] == "parent_concept_ids_flat":
                        assert any(
                            [
                                concept_filter["value"]
                                in concept.parent_concept_ids_flat
                                for concept in passage_match.concepts
                            ]
                        )
                    else:
                        assert any(
                            [
                                concept.__getattribute__(concept_filter["name"])
                                == concept_filter["value"]
                                for concept in passage_match.concepts
                            ]
                        )

    assert mock_corpora_exist_in_db.assert_called
