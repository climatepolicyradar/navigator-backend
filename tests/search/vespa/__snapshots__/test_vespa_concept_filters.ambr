# serializer version: 1
# name: test_concept_filters[-concept_filters2]
  SearchResponse(hits=1, total_family_hits=1, query_time_ms=0, total_time_ms=0, continuation_token=None, this_continuation_token='', prev_continuation_token=None, families=[SearchResponseFamily(family_slug='national-environment-policy-of-guinea_f0df', family_name='National Environment Policy of Guinea', family_description='This document notably 1) sets adaptation to climate change as a prominent goal, 2) promotes research in areas related to climate change, biological diversity and the fight against desertification, particularly in sectors related to agrofuels, non-wood forest products and forestry, and 3) promotes measures towards soil restoration and carbon sequestration.', family_category='Executive', family_date='2011-08-01T00:00:00+00:00', family_last_updated_date='2011-08-01T00:00:00+00:00', family_source='CCLW', corpus_import_id='CCLW.corpus.1.0', corpus_type_name='UNFCCC Submissions', family_geographies=['GIN'], family_metadata={'sector': ['Price', 'Government']}, family_title_match=False, family_description_match=False, total_passage_hits=1, family_documents=[SearchResponseFamilyDocument(document_title='doc name', document_slug='national-environment-policy-of-guinea_86dc', document_type='Executive', document_source_url='https://earthguinea.org/wp-content/uploads/2017/09/politique-nationale-de-l-environnement-2011-nov-2012.pdf', document_url='https://cdn.climatepolicyradar.org/navigator/GIN/2011/national-environment-policy-of-guinea_00ce569611f01abe4f7e06f40558ada8.pdf', document_content_type='application/pdf', document_passage_matches=[SearchResponseDocumentPassage(text='The involvement of local populations has led to the establishment of more than 300 forest groups in village forests over 10,000 ha and in 6 classified forests over 97,000 ha (MAEF, 2006). In addition, 185 forests in the domain of decentralized communities and totaling 38,500 ha have been classified with the assistance of the forestry administration.', text_block_id='p_60_b_6', text_block_page=61, text_block_coords=[(51.79242706298828, 318.3060302734375), (537.0337524414062, 318.3060302734375), (537.0337524414062, 393.34649658203125), (51.79242706298828, 393.34649658203125)], concepts=[Concept(id='0', name='sectors', parent_concepts=[{'name': 'Q0', 'id': '0'}, {'name': 'Q1', 'id': '1'}], parent_concept_ids_flat='Q0,Q1,', model='sectors_1', end=12, start=6, timestamp=datetime.datetime(2021, 6, 1, 0, 0, tzinfo=TzInfo(UTC)))])])], continuation_token=None, prev_continuation_token=None)])
# ---
# name: test_concept_filters[the-concept_filters0]
  SearchResponse(hits=2, total_family_hits=2, query_time_ms=0, total_time_ms=0, continuation_token=None, this_continuation_token='', prev_continuation_token=None, families=[SearchResponseFamily(family_slug='agriculture-sector-plan-2015-2019_7999', family_name='Agriculture Sector Plan 2015-2019', family_description="<br>&nbsp;<br>&nbsp;<br>&nbsp;<br>&nbsp;The Agriculture Sector Plan aims at improving food and nutritional security, develop sustainable food production and management systems, support sustainable economic growth and improve the balance of trade, strengthen natural resource management, and improve the capacity-building of stakeholders and agricultural staff. The document discusses climate change impacts and the solutions the country must implement to secure its agriculture's resilience.<br>&nbsp;<br>&nbsp;<br>&nbsp;<br><br>", family_category='Executive', family_date='2015-12-25T00:00:00+00:00', family_last_updated_date='2015-12-25T00:00:00+00:00', family_source='CCLW', corpus_import_id='CCLW.corpus.2.0', corpus_type_name='Laws and Policies', family_geographies=['NIU'], family_metadata={'topic': ['Mitigation', 'Misc'], 'sector': ['Price', 'Government'], 'keyword': ['Institutions', 'Disaster Risk Management'], 'instrument': ['Capacity building']}, family_title_match=False, family_description_match=False, total_passage_hits=1, family_documents=[SearchResponseFamilyDocument(document_title='doc name', document_slug='agriculture-sector-plan-2015-2019_16c6', document_type='Executive', document_source_url='http://www.lse.ac.uk/GranthamInstitute/wp-content/uploads/2018/04/Niue-Agriculture-Sector-Plan-2015-19.pdf', document_url='https://cdn.climatepolicyradar.org/navigator/NIU/2015/agriculture-sector-plan-2015-2019_b38dcb601d6f08bbd56f08e815256f56.pdf', document_content_type='application/pdf', document_passage_matches=[SearchResponseDocumentPassage(text='Niue imports ten times more than it exports; the import value averages NZD 4.3 million/annum and the export value averages NZD 0.27 million/annum which are often imported food. Private sector subsistence activity accounted for approximately one quarter of gross domestic product (GDP) in 2009 and subsistence agriculture and fishing remain important. The major agricultural export commodity is taro, targeting the niche markets which are mainly Niueans Citizens living in New Zealand and Australia. Other agricultural commodities exported include small amounts of honey, noni and vanilla. Niue face many challenges to meet export requirements ranging from volume to meet demand, transportation, sanitary and phytosanitary and food safety requirements and accessing export markets.', text_block_id='p_20_b_0', text_block_page=21, text_block_coords=[(55.0, 520.0), (290.0, 520.0), (290.0, 686.0), (55.0, 686.0)], concepts=[Concept(id='2', name='environment', parent_concepts=[{'name': 'Q2', 'id': '2'}, {'name': 'Q3', 'id': '3'}], parent_concept_ids_flat='Q2,Q3,', model='sectors_1', end=10, start=8, timestamp=datetime.datetime(2021, 6, 1, 0, 0, tzinfo=TzInfo(UTC)))])])], continuation_token=None, prev_continuation_token=None)])
# ---
# name: test_concept_filters[the-concept_filters1]
  SearchResponse(hits=1, total_family_hits=1, query_time_ms=0, total_time_ms=0, continuation_token=None, this_continuation_token='', prev_continuation_token=None, families=[SearchResponseFamily(family_slug='national-environment-policy-of-guinea_f0df', family_name='National Environment Policy of Guinea', family_description='This document notably 1) sets adaptation to climate change as a prominent goal, 2) promotes research in areas related to climate change, biological diversity and the fight against desertification, particularly in sectors related to agrofuels, non-wood forest products and forestry, and 3) promotes measures towards soil restoration and carbon sequestration.', family_category='Executive', family_date='2011-08-01T00:00:00+00:00', family_last_updated_date='2011-08-01T00:00:00+00:00', family_source='CCLW', corpus_import_id='CCLW.corpus.1.0', corpus_type_name='UNFCCC Submissions', family_geographies=['GIN'], family_metadata={'sector': ['Price', 'Government']}, family_title_match=False, family_description_match=False, total_passage_hits=1, family_documents=[SearchResponseFamilyDocument(document_title='doc name', document_slug='national-environment-policy-of-guinea_86dc', document_type='Executive', document_source_url='https://earthguinea.org/wp-content/uploads/2017/09/politique-nationale-de-l-environnement-2011-nov-2012.pdf', document_url='https://cdn.climatepolicyradar.org/navigator/GIN/2011/national-environment-policy-of-guinea_00ce569611f01abe4f7e06f40558ada8.pdf', document_content_type='application/pdf', document_passage_matches=[SearchResponseDocumentPassage(text='The involvement of local populations has led to the establishment of more than 300 forest groups in village forests over 10,000 ha and in 6 classified forests over 97,000 ha (MAEF, 2006). In addition, 185 forests in the domain of decentralized communities and totaling 38,500 ha have been classified with the assistance of the forestry administration.', text_block_id='p_60_b_6', text_block_page=61, text_block_coords=[(51.79242706298828, 318.3060302734375), (537.0337524414062, 318.3060302734375), (537.0337524414062, 393.34649658203125), (51.79242706298828, 393.34649658203125)], concepts=[Concept(id='0', name='sectors', parent_concepts=[{'name': 'Q0', 'id': '0'}, {'name': 'Q1', 'id': '1'}], parent_concept_ids_flat='Q0,Q1,', model='sectors_1', end=12, start=6, timestamp=datetime.datetime(2021, 6, 1, 0, 0, tzinfo=TzInfo(UTC)))])])], continuation_token=None, prev_continuation_token=None)])
# ---
